### PR TITLE
Introduce p2-pcctl

### PR DESCRIPTION
**Pull request by rudle**
Basic functionality about create, show & delete is here. Will add update in a following change

### CODE REVIEW

File: bin/p2-pcctl/main.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/square/p2/pkg/kp/pcstore"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/pc/control"
+	"github.com/square/p2/pkg/pc/fields"
+	"github.com/square/p2/pkg/types"
+)
+
+const (
+	CmdCreate = "create"
+	CmdShow   = "show"
```
Comment (mpuncel) [inline bin/p2-pcctl/main.go]: nit: this is `get` in `p2-rctl`, would be nice to standardize that

File: bin/p2-pcctl/main.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	createAZ          = cmdCreate.Flag("az", "The availability zone of the pod cluster").Required().String()
+	createName        = cmdCreate.Flag("name", "The cluster name (ie. staging, production)").Required().String()
+	createSelector    = cmdCreate.Flag("selector", "A pod selector to use for this pod cluster. This will be generated from other arguments if not provided").String()
+	createAnnotations = cmdCreate.Flag("annotations", "Complete set of annotations - must parse as JSON!").String()
+
+	cmdShow      = kingpin.Command(CmdShow, "Show a pod cluster. ")
+	showPodID    = cmdShow.Flag("pod", "The pod ID on the pod cluster").Required().String()
+	showAZ       = cmdShow.Flag("az", "The availability zone of the pod cluster").Required().String()
+	showName     = cmdShow.Flag("name", "The cluster name (ie. staging, production)").Required().String()
+	showSelector = cmdShow.Flag("selector", "A pod selector to use for this pod cluster. This will be generated from other arguments if not provided").String()
```
Comment (mpuncel) [inline bin/p2-pcctl/main.go]: is this used?

File: bin/p2-pcctl/main.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		selector := selectorFrom(az, cn, podID)
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", os.Getenv("USER")))
+		pc, err := pccontrol.Get()
+		if err != nil {
+			fmt.Printf("Caught error while fetching pod cluster: %v", err)
+			os.Exit(1)
+		}
+		fmt.Printf("Pod Cluster:\n%+v", pc)
+	case CmdDelete:
+		az := fields.AvailabilityZone(*showAZ)
```
Comment (mpuncel) [inline bin/p2-pcctl/main.go]: these should be using `deleteAZ` etc no?

File: bin/p2-pcctl/main.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	showPodID    = cmdShow.Flag("pod", "The pod ID on the pod cluster").Required().String()
+	showAZ       = cmdShow.Flag("az", "The availability zone of the pod cluster").Required().String()
+	showName     = cmdShow.Flag("name", "The cluster name (ie. staging, production)").Required().String()
+	showSelector = cmdShow.Flag("selector", "A pod selector to use for this pod cluster. This will be generated from other arguments if not provided").String()
+
+	cmdDelete      = kingpin.Command(CmdDelete, "Show a pod cluster. ")
+	deletePodID    = cmdDelete.Flag("pod", "The pod ID on the pod cluster").Required().String()
+	deleteAZ       = cmdDelete.Flag("az", "The availability zone of the pod cluster").Required().String()
+	deleteName     = cmdDelete.Flag("name", "The cluster name (ie. staging, production)").Required().String()
+	deleteSelector = cmdDelete.Flag("selector", "A pod selector to use for this pod cluster. This will be generated from other arguments if not provided").String()
```
Comment (mpuncel) [inline bin/p2-pcctl/main.go]: don't think you're gonna need this either

File: bin/p2-pcctl/main.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	logger := logging.NewLogger(logrus.Fields{})
+	pcstore := pcstore.NewConsul(client, 3, &logger)
+
+	switch cmd {
+	case CmdCreate:
+		az := fields.AvailabilityZone(*createAZ)
+		cn := fields.ClusterName(*createName)
+		podID := types.PodID(*createPodID)
+		selector := selectorFrom(az, cn, podID)
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", os.Getenv("USER")))
```
Comment (mpuncel) [inline bin/p2-pcctl/main.go]: is `USER` an environment variable we can expect to be available on all systems?

File: bin/p2-pcctl/main.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		cn := fields.ClusterName(*showName)
+		podID := types.PodID(*showPodID)
+		selector := selectorFrom(az, cn, podID)
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", os.Getenv("USER")))
+		pc, err := pccontrol.Get()
+		if err != nil {
+			fmt.Printf("Caught error while fetching pod cluster: %v", err)
+			os.Exit(1)
+		}
+		fmt.Printf("Pod Cluster:\n%+v", pc)
```
Comment (mpuncel) [inline bin/p2-pcctl/main.go]: you wanted to remove this so the output is pure json

File: bin/p2-pcctl/main.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	case CmdDelete:
+		az := fields.AvailabilityZone(*showAZ)
+		cn := fields.ClusterName(*showName)
+		podID := types.PodID(*showPodID)
+		selector := selectorFrom(az, cn, podID)
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", os.Getenv("USER")))
+		pccontrol.Delete()
```
Comment (anthonybishopric) [inline bin/p2-pcctl/main.go]: Err check here?

File: bin/p2-pcctl/main.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	case CmdShow:
+		az := fields.AvailabilityZone(*showAZ)
+		cn := fields.ClusterName(*showName)
+		podID := types.PodID(*showPodID)
+		selector := selectorFrom(az, cn, podID)
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", os.Getenv("USER")))
+		pc, err := pccontrol.Get()
+		if err != nil {
+			fmt.Printf("Caught error while fetching pod cluster: %v", err)
+			os.Exit(1)
```
Comment (anthonybishopric) [inline bin/p2-pcctl/main.go]: `log.Fatalf` to be consistent with above

File: bin/p2-pcctl/main.go (hunk 2 above)
Comment (rudle) [inline bin/p2-pcctl/main.go]: No, it's theoretical at this point. I'm going to remove it until there's a concrete motivation for it.

Aside: is there an convention in open source for user-specified selectors as strings? I'd prefer to re-use a parser to implementing one.

Comment (mpuncel) [inline bin/p2-pcctl/main.go]: I think we usually use `klabels.Parse()`, let me dig around

**Comment by rudle** [comment]
All comments have been addressed. The tests are new since last review, PTAL at those.

File: bin/p2-pcctl/main.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", currentUserName()))
+		pc, err := pccontrol.Get()
+		if err != nil {
+			fmt.Printf("Caught error while fetching pod cluster: %v", err)
+			os.Exit(1)
+		}
+
+		bytes, err := json.Marshal(pc)
+		if err != nil {
+			logger.WithError(err).Errorln("Unable to marshal PC as JSON")
```
Comment (anthonybishopric) [inline bin/p2-pcctl/main.go]: Should this be `Fatalf`?

File: bin/p2-pcctl/main.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+		az := fields.AvailabilityZone(*getAZ)
+		cn := fields.ClusterName(*getName)
+		podID := types.PodID(*getPodID)
+		selector := selectorFrom(az, cn, podID)
+
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", currentUserName()))
+		pc, err := pccontrol.Get()
+		if err != nil {
+			fmt.Printf("Caught error while fetching pod cluster: %v", err)
+			os.Exit(1)
```
Comment (anthonybishopric) [inline bin/p2-pcctl/main.go]: Nit: everywhere else seems to use `log.Fatalf`

File: bin/p2-pcctl/main.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+		}
+		fmt.Printf("%s", bytes)
+	case CmdDelete:
+		az := fields.AvailabilityZone(*deleteAZ)
+		cn := fields.ClusterName(*deleteName)
+		podID := types.PodID(*deletePodID)
+		selector := selectorFrom(az, cn, podID)
+		pccontrol := control.NewPodCluster(az, cn, podID, pcstore, selector, kv.NewUnmanagedSession("pcctl", currentUserName()))
+		errors := pccontrol.Delete()
+		if len(errors) > 1 {
```
Comment (anthonybishopric) [inline bin/p2-pcctl/main.go]: Check is wrong (should be `>=1`) and also is unnecessary. Just delete it.

File: bin/p2-pcctl/main.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+			for _, err := range errors {
+				_, _ = os.Stderr.Write([]byte(fmt.Sprintf("Failed to one pod cluster matching arguments. Error:\n %s\n", err.Error())))
+			}
```
Comment (anthonybishopric) [inline bin/p2-pcctl/main.go]: It'd be a desirable property to exit non-0 if there are any errors (so maybe keep the if statement above, just fix it)

**Comment by anthonybishopric** [comment]
🚢

**Pull request merged**